sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveExternalCatalog.scala

```diff
@@ -810,13 +840,37 @@ private[spark] class HiveExternalCatalog(conf: SparkConf, hadoopConf: Configurat
       newSpecs: Seq[TablePartitionSpec]): Unit = withClient {
     client.renamePartitions(
       db, table, specs.map(lowerCasePartitionSpec), newSpecs.map(lowerCasePartitionSpec))
+
+    val tableMeta = getTable(db, table)
+    val partitionColumnNames = tableMeta.partitionColumnNames
+    // Hive metastore is not case preserving and keeps partition columns with lower cased names.
+    // When Hive rename partition for managed tables, it will update the partition location with
+    // a default value generate by the new spec and lower cased partition column names. This is
+    // unexpected and we need to call `alterPartitions` to fix the partition location.
+    val hasUpperCasePartitionColumn = partitionColumnNames.exists(col => col.toLowerCase != col)
+    if (tableMeta.tableType == MANAGED && hasUpperCasePartitionColumn) {
+      val tableLocation = tableMeta.storage.locationUri.get
+      val newParts = newSpecs.map { spec =>
+        val partition = client.getPartition(db, table, lowerCasePartitionSpec(spec))
+        val partitionPath = new Path(
+          tableLocation,
+          partitionColumnNames.map(col => col + "=" + spec(col)).mkString("/"))
```

Comment:
RENAME PARTITION can only change the partition spec, not partition path. The SQL syntax is :`ALTER TABLE table_name PARTITION partition_spec RENAME TO PARTITION partition_spec;`, users can't specify partition location